File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int ret;
int N;
vector<vector<int>> v;
int visited[10];

void DFS_recur(int curr,int k) {
    ret=max(ret,curr);
    if(curr==N) {
        return;
    }
    // cout << curr << "\n";
    
    for(int i=0;i<N;i++) {
        if(visited[i]==0 and k>=v[i][0]) {
            visited[i]=1;
            DFS_recur(curr+1,k-v[i][1]);
            visited[i]=0;
        }
    }
}
int solution(int k, vector<vector<int>> dungeons) {
    N=dungeons.size();
    v=dungeons;
    
    DFS_recur(0,k);
    return ret;
}

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    
    static int M;
    static int ret;
    static int visited[] = new int[10];
    static int temp[][];
    
    static void DFS_recur(int curr,int health) {
        
        ret=Math.max(ret,curr);
        
        for(int i=0;i<M;i++) {
            if(visited[i]==0 && temp[i][0]<=health) {
                visited[i]=1;
                DFS_recur(curr+1,health-temp[i][1]);
                visited[i]=0;
            }
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        
        
        M=dungeons.length;
        temp=dungeons;
        
        DFS_recur(0,k);
        
        return ret;
    }
}